# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/jit-test/tests/typedarray/ensure-non-inline.js
# Commit: e39ae7aad8d9
# Full Hash: e39ae7aad8d93497b7a7e6ccf4ee675b4631c28e
# Author: Steve Fink <sfink@mozilla.com>
# Date: 2023-09-21 03:21:32
# Regressor Bug: 1690111
# File Overlap Count: 2
# Description:
#   Bug 1690111 - Implement JS::EnsureNonInlineArrayBufferOrView. r=jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D176596
# ==============================================================================

diff -r 603e8067e68f -r e39ae7aad8d9 js/src/jit-test/tests/typedarray/ensure-non-inline.js
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/js/src/jit-test/tests/typedarray/ensure-non-inline.js	Wed Sep 20 09:42:00 2023 +0000
@@ -0,0 +1,78 @@
+const constructors = [
+    Int8Array,
+    Uint8Array,
+    Uint8ClampedArray,
+    Int16Array,
+    Uint16Array,
+    Int32Array,
+    Uint32Array,
+    Float32Array,
+    Float64Array ];
+
+function messWith(view, obj) {
+    view[0] = 1;
+    view[1] = 1;
+    view[2] = 2;
+    view[3] = 3;
+    ensureNonInline(obj);
+    assertEq(view[0], 1);
+    assertEq(view[1], 1);
+    assertEq(view[2], 2);
+    assertEq(view[3], 3);
+}
+
+function test() {
+    // Bufferless
+    for (const ctor of constructors) {
+        let small = new ctor(4);
+        messWith(small, small);
+        let big = new ctor(4000);
+        messWith(big, big);
+    }
+
+    // With buffer, single view, operate on view
+    for (const ctor of constructors) {
+        let ab = new ArrayBuffer(96);
+        const small = new ctor(ab);
+        messWith(small, small);
+        ab = new ArrayBuffer(4000);
+        const big = new ctor(ab);
+        messWith(big, big);
+    }
+
+    // With buffer, single view, operate on ArrayBuffer
+    for (const ctor of constructors) {
+        let ab = new ArrayBuffer(96);
+        const small = new ctor(ab);
+        messWith(small, ab);
+        ab = new ArrayBuffer(4000);
+        const big = new ctor(ab);
+        messWith(big, ab);
+    }
+
+    // With buffer, dual view, operate on view
+    for (const ctor of constructors) {
+        let ab = new ArrayBuffer(96);
+        let view0 = new Uint8Array(ab);
+        const small = new ctor(ab);
+        messWith(small, small);
+        ab = new ArrayBuffer(4000);
+        view0 = new Uint8Array(ab);
+        const big = new ctor(ab);
+        messWith(big, big);
+    }
+
+    // With buffer, dual view, operate on ArrayBuffer
+    for (const ctor of constructors) {
+        let ab = new ArrayBuffer(96);
+        let view0 = new Uint8Array(ab);
+        const small = new ctor(ab);
+        messWith(small, ab);
+        ab = new ArrayBuffer(4000);
+        view0 = new Uint8Array(ab);
+        const big = new ctor(ab);
+        messWith(big, ab);
+    }
+}
+
+test();